### PR TITLE
[react] allow weaker prop declarations for nullish properties

### PR DESCRIPTION
**Pull request by ferdaber**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/28015#issuecomment-446310610
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

cc @andrewbranch

### CODE REVIEW

File: types/prop-types/index.d.ts [hunk 1]
```diff
@@ -5,11 +5,15 @@
...
 // TypeScript Version: 2.8
 
+export type ReactComponentLike =
+    | string
+    | ((props: any, context?: any) => null)
+    | (new (props: any, context?: any) => null);
+
 export interface ReactElementLike {
-    type: string | ((...args: any[]) => ReactElementLike);
+    type: ReactComponentLike;
```
Comment (ferdaber) [inline types/prop-types/index.d.ts]: The original `type` property for this was not exactly the same as the one in the React type definitions, causing issues with the `PropTypes.element` validator, which this fixes.

File: types/prop-types/prop-types-tests.ts [hunk 2]
```diff
@@ -129,34 +129,34 @@ type ExtractedPartialProps = PropTypes.InferProps<typeof partialPropTypes>;
 type ExtractedPropsWithoutAnnotation = PropTypes.InferProps<typeof propTypesWithoutAnnotation>;
 type ExtractedPropsFromOuterPropsWithoutAnnotation = PropTypes.InferProps<typeof outerPropTypesWithoutAnnotation>['props'];
 
-// $ExpectType: true
+// $ExpectType true
```
Comment (ferdaber) [inline types/prop-types/prop-types-tests.ts]: Fixed a longstanding issue with these DTSLint assertions

File: types/react/index.d.ts [hunk 3]
```diff
@@ -2558,6 +2558,14 @@ declare namespace React {
 
     type ValidationMap<T> = PropTypes.ValidationMap<T>;
 
+    type WeakValidationMap<T> = {
```
Comment (ferdaber) [inline types/react/index.d.ts]: The weak validation map will transform any prop declaration that can be `null` or `undefined` and widen it to be both: `null | undefined`. This allows props authors to declare something like this:
```ts
type Props = { foo?: string }
```

and have it pass checks for HOCs that require it to be:
```ts
type Props = { foo?: string | null }
```

Most developers do not realize that `propTypes` allows for both so this should meet a compromise of DX and runtime accuracy.

**Comment by johnnyreilly** [comment]
Thanks!

**Comment by milesj** [comment]
Would this satisfy the commented out tests here? https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/test/managedAttributes.tsx

**Comment by Kovensky** [comment]
Unfortunately no, it requires bumping `@types/react` to TS 3.0.

**Comment by ferdaber** [comment]
@milesj if you were asking if those tests passed, I did test them out and they did not error for TS 3.0.

File: types/react/index.d.ts (hunk 3 above)
Comment (chibicode) [reply types/react/index.d.ts]: Hi @ferdaber, after upgrading `@types/react` to 16.7.14 from 16.7.13, my project is getting

```
TS2605: JSX element type 'ReactElement<any> | null' is not a constructor function for JSX elements.
  Type 'ReactElement<any>' is not assignable to type 'Element'.
```

for almost every component. I've tracked down this bug into these changes. I'm wondering if anyone else is getting this error...I'll try to create a reproducible repo soon.

Comment (chibicode) [reply types/react/index.d.ts]: Actually, the problem was a different issue (cc @cuyl) - I've written the fix here:

https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17161#issuecomment-446933731

Comment (cuyl) [reply types/react/index.d.ts]: thx

**Pull request merged**